dashboard/src/modules/components/HeaderComponent/index.jsx

```diff
@@ -34,14 +34,19 @@ import { useLocation, useNavigate } from "react-router-dom";
 import Cookies from "js-cookie";
 import { logout } from "actions/authActions";
 import pbenchLogo from "assets/logo/pbench_logo.svg";
+import { useKeycloak } from '@react-keycloak/web';
+import { movePage } from "actions/authActions";
 
 const HeaderToolbar = () => {
   const dispatch = useDispatch();
-  const loginDetails = useSelector((state) => state.userAuth.loginDetails);
+  const username = Cookies.get("username");
```

Comment:
Just out of curiosity, why are we keeping the username in a cookie instead of in the Redux store?

Reply:
If the page is reloaded then the user wouldn't have to re-login, the keycloak object checks for the active SSO sessions of the user and if there is an active session then the user is logged in. 

And actually you right we don't need to store the username in the cookie or redux the store at all. We can always get the username by referencing the `keycloak` object directly like `keycloak.tokenParsed?.preferred_username` 

> I don't know how much control we have over the tokens, now that we're using the Keycloak plug-in -- how will the plug-in respond to the page being reloaded? (I.e., how does _it_ store stuff??)

The keycloak object will automatically refresh the tokens in the background whenever a token is near expiry so when we use `keycloak.token` of an active keycloak object we should always have a valid token.

Reply:
> If the page is reloaded then [...] the keycloak object checks for the active SSO sessions

So, then, the  keyclock object must be storing the session ID and at least the refresh token in a cookie somewhere, eh?

> we don't need to store the username in the cookie or redux the store at all

😎 

> we should always have a valid token

Only if keycloak is clever enough to survive a reload!...  😉

Reply:
> So, then, the keyclock object must be storing the session ID and at least the refresh token in a cookie somewhere, eh?

We can get the current refresh token by accessing `keycloak.refreshToken` and get the current access token by accessing `keycloak.token` and sessionID as `keycloak.sessionId` https://www.npmjs.com/package/keycloak-js?activeTab=code


> Only if keycloak is clever enough to survive a reload!

yeah, we could store the keycloak object in the browser's local storage but that may not be very ideal and need some research on that.

Reply:
> We can get the current refresh token by accessing `keycloak.refreshToken` and get the current access token by accessing `keycloak.token` and sessionID as `keycloak.sessionId`

Yeah, but after a reload, where do we get the `keycloak` object from and how do those values get repopulated?

> we could store the keycloak object

This is my question:  do we need to protect Keycloak, or will it protect itself?  (I looked at some of the code that you pointed to, and I see it has support for manipulating cookies, but I didn't find where it actually creates any.)

Reply:
> Yeah, but after a reload, where do we get the `keycloak` object from and how do those values get repopulated?

I am not entirely sure but I think the `keycloak` object is stored in the browser's memory but I am also storing it as our redux state. All the properties of the keycloak object gets repopulated after the App `check-sso` is complete on line 75 of App.js.


> This is my question: do we need to protect Keycloak, or will it protect itself? (I looked at some of the code that you pointed to, and I see it has support for manipulating cookies, but I didn't find where it actually creates any.)

I don't think it stores any cookies by itself, a new object is created upon each reload and it is stored in a similar way as redux state.

Reply:
[The docs](https://www.keycloak.org/docs/latest/securing_apps/index.html#session-status-iframe) seem to indicate that the Keycloak plugin maintains a cookie with the login status...which would explain how reloading the page does not result in a logout.

Between the fact that we configure the Keycloak plugin when we (re)load our app (via [our reference](https://github.com/distributed-system-analysis/pbench/blob/d2f5799c4f91a04ef68f38ed37cded380d2f9cc9/dashboard/src/App.js#L71) to the the `<ReactKeycloakProvider>` tag) and the fact that the plugin maintains an iframe and cookies, I don't think we need to store our `keycloak` object in persistent storage:  if our app is reloaded, we'll just recreate the object.  (This seems like a Good Thing™ to me.)